Feature/search user in directory

sppanday opened the pull request:
This pull request brings changes for role directory and partially touches on services and people directory. This will automatically leverage work towards service directory and people directory at a portion.

194064: Web: As an Authenticated Practitioner, I should be able to Search the Role Directory, so that I can locate a Role and view their details.

224274: As an authenticated Practitioner I would like to see my first and last name initials on avatar.

224329: Web: As an Authenticated Practitioner, all avatars on web display Lingo colors by role category so that I am shown accessibility friendly colors.

sppanday committed:
implemented role directory detail page view

sppanday committed:
translate roles with enums

sppanday committed:
config values for roles and service directory

sppanday committed:
directory view cleanup

sppanday committed:
aligned data and cleaned up role detail view

sppanday committed:
moved table logic separately for better separation

sppanday committed:
separated role detial api call

sppanday committed:
translated value field with enum

sppanday committed:
removed duplicate data, enum cleanup

sppanday committed:
method name change

sppanday committed:
file name changed for role detail view

sppanday committed:
implemented first,last name on avatar

sppanday committed:
role drop down on left panel

sppanday committed:
feature documentation for role directory

sppanday committed:
text cleanup

sppanday committed:
logic to display role directory dialog

sppanday committed:
logic to trigger role search dialog to apear

sppanday committed:
logic to show drop down for roles on left

sppanday committed:
removed unused function on role detail view

sppanday committed:
configs for role search

sppanday committed:
removing translation for role directory drop down

sppanday committed:
removed bunch of console.log

sppanday committed:
Directory - role search functionality

sppanday committed:
config item for avatar color

sppanday committed:
made sorting to be alphabetically via config

sppanday committed:
feature documentation for role directory search

sppanday committed:
minor changes on Filter by favorite events

sppanday committed:
readjusted role detail fetch function

sppanday committed:
enhanced documentation for role directory search

sppanday committed:
Finalised avatar color with configurable parameter

sppanday committed:
change 'invite' to 'discussion'

sppanday committed:
fixed bug on filled status not updating correctly

sppanday committed:
Added practitioners fulfilling role

sppanday committed:
indexing in view for role fulfilling

sppanday committed:
pointed role detial api to right api

sppanday committed:
added api prefix

sppanday committed:
re ordered render method, variables renamed

sppanday committed:
get rid of extra space

sppanday committed:
cleanup on role text

sppanday committed:
added role icon in homepage tile

sppanday committed:
added default to avatar colors not found in config

sppanday committed:
improved conditional checking for role search

sppanday committed:
Better way to find name from given matrix id

sppanday committed:
simplified avatar to have first/last name initials

sppanday committed:
refactored avatar name/id matching scenario

sppanday committed:
Changed text on left panel as we dont support
explore public room so text has been rewored

sppanday committed:
included null/undefined check

sppanday committed:
bunch of refactor on role directory for default

sppanday committed:
simplified directory search boolean check

jrschremmer committed:
Merge branch 'master' of https://github.com/fhirfactory/pegacorn-communicate-matrix-react-sdk into feature/search-user-in-directory

# Conflicts:
#	feature-communicate-list.md
#	src/components/structures/HomePage.tsx
#	src/config.ts

jrschremmer commented on src/Avatar.ts:
```diff
@@ -97,7 +98,7 @@ export function defaultAvatarUrlForString(s: string): string {
     const cssVariable = `--avatar-background-colors_${colorIndex}`;
     const cssValue = document.body.style.getPropertyValue(cssVariable);
-    const color = cssValue || defaultColors[colorIndex];
+    const color = ((config.avatarColors) ? getSpecificColorFromConfig(s) : undefined) || cssValue || defaultColors[colorIndex];
```
Please insert a new line to make the comparison to existing element code easier e.g.
```
    const color = ((config.avatarColors) ? getSpecificColorFromConfig(s) : undefined) || 
                  cssValue || defaultColors[colorIndex];
```

sppanday replied:
No worries change has been implemented exactly as above in new line.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +113,31 @@ export function defaultAvatarUrlForString(s: string): string {
+ * Not having this config or not having color in config parameter would default
+ * avatar colors back to matrix color
+ * @param {string} s
+ * @param {string} randomlySelectedAvatarColor randomly selected color from config
```
Please remove these 3 params in the code comment as they aren't parameters passed in

sppanday replied:
These have been removed and a bit of more detailed explanation has been added above these lines for clarity.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +113,31 @@ export function defaultAvatarUrlForString(s: string): string {
+    const colorIndex = total % numOfColors;
+    const defaultToOtherColors = String(Object.values(configAvatarColors)[colorIndex]);
+    let randomlySelectedAvatarColor = String(Object.values(configAvatarColors)[Math.floor(Math.random() * numOfColors)]);
+    return (avatarColors || defaultToOtherColors || randomlySelectedAvatarColor);
```
I'd suggest refactoring this, so as soon as you have a result return it, instead of calculating defaultToOtherColors and randomlySelectedAvatarColor unnecessarily

sppanday replied:
Change has been implemented in new commit. I have refactored to address suggested change "refactoring this, so as soon as you have a result return it, instead of calculating defaultToOtherColors and randomlySelectedAvatarColor"

jrschremmer replied:
Has this been pushed as I couldn't see it?

sppanday replied:
I refactored and ...I pushed following yesterday:

   `let randomlySelectedAvatarConfigColor = String(Object.values(configAvatarColors)[Math.floor(Math.random() * numOfColors)]);`
   ` const color = avatarFromConfigColors || defaultToOtherConfigColors || randomlySelectedAvatarConfigColor;`
   ` return color;`

sppanday replied:
Disregard above comment please. Please refer to latest pushes , i have removed both variable , retained defaultToOtherColors logic in different way then removed randomColorSelection entirely as it was not reliable strategy.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +113,31 @@ export function defaultAvatarUrlForString(s: string): string {
+    let configAvatarColors = config.avatarColors;
+    let avatarColors = configAvatarColors[s];
+    let numOfColors = Object.keys(configAvatarColors).length;
+    let total = 0;
```
Please add a code comment describing the purpose of the defaultToOtherColors logic

sppanday replied:
This has been actioned with a good chunk of description for defaultToOtherColors and function name has been slightly renamed as well with new commit

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +113,31 @@ export function defaultAvatarUrlForString(s: string): string {
+    }
+    const colorIndex = total % numOfColors;
+    const defaultToOtherColors = String(Object.values(configAvatarColors)[colorIndex]);
+    let randomlySelectedAvatarColor = String(Object.values(configAvatarColors)[Math.floor(Math.random() * numOfColors)]);
```
Under what circumstances is defaultToOtherColors nullish?  As it is not clear when randomlySelectedAvatarColor would ever be used.

sppanday replied:
It captures both nullish and undefined results and goes to next one but yeah chance of randomlySelected color being used are based on error in calculating text. But it is a rare case of 5-10% errored cases of name not being passed in function when. These may include:
1) System failure where name is not string or name is passed as null into this calculator.
2)  Name is not in valid text format 
3) Developer changing something erratically on function that sends name into this function as parameter, then it cant resolve name as string so returns undefind/null, and in that case use randomly selected color instead of using matrix green colors.
These would result final way of going to last one as randomly selected color.
**Please find my new changes pushed for requested changes as well for previous comments in new commits**

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +214,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+
+        // find out if name input is an email for avatar
+        const nameIsInMatrixIdFormat = name.indexOf(':') > -1;
+        if (name.length > 1 && nameIsInMatrixIdFormat) {
```
Please remove name.length > 1 && , as you already have the name.length == 1 check above

sppanday replied:
Done.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +214,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+            //find first and last names from email address
+            let nameFromMatrixId = name.split(':')[0].split('.');
+            firstName = nameFromMatrixId[0];
+            if (nameFromMatrixId[nameFromMatrixId.length - 1]) {
```
Why does the if logic and the then logic not align?  I'd suggest the if could be:
```
    if (nameFromMatrixId.length > 1) {
```

sppanday replied:
Done

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +214,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+        }
+            /**
+             * Split first and last-names based on regular expression. It checks following usernames patterns
+             * "firstname.lastname", "firstname-lastname", "firstname lastname, "firstname, lastname"
```
Please change "firstname, lastname" to "lastname, firstname" to match the implementation.

sppanday replied:
Done

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +214,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+        return undefined;
+    }
+
+    if (name.length == 1) {
```
To centralise logic I'd suggest the following impl:
```
    try {
        // if the name is a matrix id, remove the matrix domain
        const nameWithoutMatrixDomain = name.split(':')[0];

        /**
          * Split first and last-names based on regular expression. It checks following usernames patterns
          * "firstname.lastname", "firstname-lastname", "firstname lastname, "lastname, firstname"
          * Based on https://stackoverflow.com/questions/10346722/how-can-i-split-a-javascript-string-by-white-space-or-comma
          */
        const fullName = nameWithoutMatrixDomain.split(/[ .,-]+/);
        let firstName = fullName[0];
        let lastName = '';
        if (fullName.length > 1) { 
            lastName = fullName[1];
            if (nameWithOutMatrixDomain.includes(',')){
                firstName = fullName[1];
                lastName = fullName[0];
            }
        }
        const firstAndLastInitials = getInitialLetter(firstName) + getInitialLetter(lastName);
        return firstAndLastInitials;
```

jrschremmer replied:
Please review this suggestion

sppanday replied:
Done. Sounds good. Lets see in future if it raises any issues, for now I have replaced try block with your code.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +214,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+         * If there is a problem translating then just grab first name initials as matrix default
+         */
+        if (e instanceof Error) {
+            getInitialLetter(name);
```
This is missing the return keyword

jrschremmer replied:
Please review this suggestion

sppanday replied:
I thought yes I forgot that one but no that return is not required as the fuction being called already executes return . But i will use "return" anyways as its more safer to use it than not.

jrschremmer commented on src/PageTypes.js:
```diff
@@ -23,4 +23,5 @@ export default {
     UserView: "user_view",
     GroupView: "group_view",
     MyGroups: "my_groups",
+    RoleView: "role_directory"
```
To be consistent with the existing code I;d suggest either:
    RoleView: "role_view"
OR
    RoleDirectory: "role_directory"

jrschremmer replied:
Please review this suggestion

sppanday replied:
Earlier implementation required me to use these page types but was no longer required so now completely removed.

jrschremmer commented on src/components/structures/HomePage.tsx:
```diff
@@ -34,6 +34,7 @@ import MiniAvatarUploader, {AVATAR_SIZE} from "../views/elements/MiniAvatarUploa
 import Analytics from "../../Analytics";
 import CountlyAnalytics from "../../CountlyAnalytics";
 import * as customConfig from "../../config";
+import { role_directory_feature_name, showExplorePublicRoom, showRoleDirectory} from "../../config";
```
To be consistent with existing code, I'd suggest not importing these directly, but instead just refer to customConfig
e.g. customConfig.role_directory_feature_name

sppanday replied:
Done

jrschremmer commented on src/components/structures/HomePage.tsx:
```diff
@@ -133,13 +139,23 @@ const HomePage: React.FC<IProps> = ({ justRegistered = false }) => {
                 <AccessibleButton onClick={onClickSendDm} className="mx_HomePage_button_sendDm">
                     { _t("Send a Direct Message") }
                 </AccessibleButton>
+                {showExplorePublicRoom &&
```
Please call this showExplorePublicRooms to match the label

sppanday replied:
Done

jrschremmer commented on src/components/structures/HomePage.tsx:
```diff
@@ -133,13 +139,23 @@ const HomePage: React.FC<IProps> = ({ justRegistered = false }) => {
                     { _t("Create a Group Chat") }
                 </AccessibleButton>
             </div>
+            <div className="mx_HomePage_default_buttons">
```
Why does this need a new div?  The mx_HomePage_default_buttons should be responsive and handle a variable number of tiles

sppanday replied:
Good question. Answer is to apply style as horizontal tile from left to right it needs to be wrapped with that className and it cant go with with other ones above this as I want Service, people and Role search into new line as second row of tile so in service and people directory you will see all aligned and wrapped inside this div which is nice as directory stuff are all aligned in new line, I will show it on tuesday demo. So 3 tiles for directories are part of this div in new changes and will align nicely.

jrschremmer commented on src/components/structures/HomePage.tsx:
```diff
@@ -133,13 +139,23 @@ const HomePage: React.FC<IProps> = ({ justRegistered = false }) => {
             </div>
+            <div className="mx_HomePage_default_buttons">
+                {showRoleDirectory &&
+                    <AccessibleButton onClick={onClickRoleDirectory} className="mx_HomePage_button_role">
```
I'd suggest this should be the first tile to encourage role discussions instead of direct discussions

sppanday replied:
As I have mentioned In other thread, I have put service, people and role in second row with order of role, people, service from left to right in Line 2 horizontally. And, Send Direct message and Create a group chat are in first one. If one more of new one in future would be added , then that will be third on first line of tile.

jrschremmer commented on src/components/views/avatars/BaseAvatar.tsx:
```diff
@@ -108,7 +109,7 @@ const BaseAvatar = (props: IProps) => {
 
     if (!imageUrl && defaultToInitialLetter) {
-        const initialLetter = AvatarLogic.getInitialLetter(name);
+        const initialLetter = showFirstLastNameIntialsOnAvatarBackground ? AvatarLogic.getFirstAndLastNameInitialLetters(name) : AvatarLogic.getInitialLetter(name);
```
Please insert a new line to make the comparison to existing element code easier e.g.
```
        const initialLetter = showFirstLastNameIntialsOnAvatarBackground ? AvatarLogic.getFirstAndLastNameInitialLetters(name) : 
                              AvatarLogic.getInitialLetter(name);
```

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -42,15 +42,22 @@ import {UIFeature} from "../../../settings/UIFeature";
 export const KIND_INVITE = "invite";
 export const KIND_CALL_TRANSFER = "call_transfer";
+// role and service directory
+export const KIND_Role_Directory_Search = "search_role_directory";
```
Please make all capitals to be consistent with the existing code

sppanday replied:
Done.
Followings have been capitalized and used now.
KIND_ROLE_DIRECTORY_SEARCH,
KIND_PEOPLE_DIRECTORY_SEARCH,
KIND_SERVICE_DIRECTORY_SEARCH

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -42,15 +42,22 @@ import {UIFeature} from "../../../settings/UIFeature";
+export const KIND_People_Directory_Search = "search_people_directory";
 
-const INITIAL_ROOMS_SHOWN = 3; // Number of rooms to show at first
+const INITIAL_ROOMS_SHOWN = config.numberOfRecordsToShowInSearch || 3; // Number of rooms to show at first
```
Please insert a new line to make the comparison to existing element code easier e.g.
```
const INITIAL_ROOMS_SHOWN = config.numberOfRecordsToShowInSearch
                            || 3; // Number of rooms to show at first
```

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -162,8 +206,8 @@ class DMUserTile extends React.PureComponent<IDMUserTileProps> {
                     avatarSize, avatarSize, "crop")}
-                name={this.props.member.name}
-                idName={this.props.member.userId}
+                name={(config.avatarColors ? this.props.member.roleCategoryId: null) || this.props.member.name}
```
Please insert a new line into both of these lines to make the comparison to existing element code easier

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -246,9 +314,22 @@ class DMRoomTile extends React.PureComponent<IDMRoomTileProps> {
         return result;
     }
 
+    searchIsOnRolePeopleServiceDirectory = () => {
```
Please rename to searchIsOnRoleOrPeopleOrServiceDirectory

sppanday replied:
Done.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -264,8 +345,8 @@ class DMRoomTile extends React.PureComponent<IDMRoomTileProps> {
                     avatarSize, avatarSize, "crop")}
-                name={this.props.member.name}
-                idName={this.props.member.userId}
+                name={(config.avatarColors ? this.props.member.roleCategoryId: null) || this.props.member.name}
```
Please insert a new line into both of these lines to make the comparison to existing element code easier

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -595,7 +707,10 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
         if (this.state.filterText.startsWith('@')) {
             // Assume mxid
-            newMember = new DirectoryMember({user_id: this.state.filterText, display_name: null, avatar_url: null});
+            newMember = new DirectoryMember({
```
Please combine these 2 lines to make the comparison to existing element code easier e.g.:
```
            newMember = new DirectoryMember({user_id: this.state.filterText, display_name: null, avatar_url: null
                , available: false, favorite: false, roleCategoryId: null
            });
```

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+        ev.stopPropagation();
+        if (!this.searchIsOnRolePeopleServiceDirectory()) return;
+        const eventTarget = ev.currentTarget;
+        let filterByFavoriteEvent = eventTarget.parentNode.textContent.includes("Favorite");
```
Please move this line down to when it is required, to avoid running it unnecessarily

sppanday replied:
Done.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+        });
+    }
+
+    searchIsOnRolePeopleServiceDirectory = () => {
```
This method is duplicated please remove

sppanday replied:
Methods are in two different classes .  making them one would require a bit of work I think. It will be handled in better way in people and service search functionalities.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+                this.setState({
+                    favorites: response.favourites
+                })
+            })
```
Please handle gracefully if an error occurs e.g.
```
        }).catch(err => { 
            // show the user an error has occured
        })
```

sppanday replied:
Done. Yep, forgot to include catch block there. Thank you for picking this one.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+            this.setState({
+                numOfRecordsFromSearchAPI: totalRecords
+            });
+        })
```
Please handle gracefully if an error occurs

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+                    role_category = value["primaryRoleCategoryID"];
+                    // filled or not filled status
+                    // if activePractitionerSet array is non-empty someone is fulfilling that role
+                    if (value.activePractitionerSet?.length >= 1) {
```
this if statement can become:
```
   roleIsActive = (value.activePractitionerSet?.length >= 1);
```

sppanday replied:
Done, but I would prefer my original approach for readability, anyways  change has been done.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+                    } else {
+                        roleIsActive = false;
+                    }
+                });
```
Please handle gracefully if an error occurs

sppanday replied:
Done

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -751,7 +866,179 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
+                    });
+                }
+            }).catch(e => {
+                console.error("Error searching role directory:");
```
I think we need to show something in the browser.

sppanday replied:
Thats a valid point. But, matrix default did not show any error on UI for search on prmpt. But, it can be added if you recommend so, I think it is going to require me to add one more variable which is going to carry error. Do you recommend including this in scope of my change or leave it? If there is search error then I beleive no result would be shown but yeah it can be added if you recommend too. I will look into including error text on UI.

sppanday replied:
Displaying error if error occurs in search has been implemented and tested and pushed.

jrschremmer replied:
What does `errorText: err` end up showing in the UI?  We want to show a nice message to the user and not the raw javascript error.

sppanday replied:
Yeah it would have displayed actual error which would not make sense to end user so I have displayed nice error accordingly with later commit instead of displaying errorText as error.

sppanday replied:
Update: I have updated error updates from all catch blocks to plain english format to be consistant with matrix code.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -76,18 +83,43 @@ class Member {
+    _roleCategoryId: string;
 
-    constructor(userDirResult: {user_id: string, display_name: string, avatar_url: string}) {
+    constructor(userDirResult: {user_id: string, display_name: string, avatar_url: string, favorite: boolean, available: boolean, roleCategoryId: string}) {
```
Can we overload the constructor, so the defaults of false, false, null can be centralised?

sppanday replied:
The value has been set to false in constructor and made optional for string as well and also put my changes in new line which I should have done earlier. Using question mark '?' will make it optional and earlier befor eyou had pointed it out, I had tested with matrix default dialog window where invit is sent and it works well so where objec tmember is called, null default is set for string and false for booleans. But with your recommendation I made one more final change to make these values optional in constructor and tested it out again. Works well again with matrix default send invite to user window which opens from left plus menu. I have continued to use this window until i finish directory work so that I can test if default behavior is still kept same by making my implementations optional fields. Based on  my bunch of previous refactors it has shown that using question mark is making values optional wherever it is used.

jrschremmer replied:
Should roleCategoryId also have the question mark?

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -1099,6 +1413,10 @@ export default class InviteDialog extends React.PureComponent<IInviteDialogProps
                 onToggle={this._toggleMember}
                 highlightWord={this.state.filterText}
                 isSelected={this.state.targets.some(t => t.userId === r.userId)}
+                isFavorite={(this.state.favorites?.indexOf(r.user.name)  !== -1 ? true: false) || false}
```
Please remove 
? true: false 
as it is redundant

sppanday replied:
This one has been actioned.

jrschremmer commented on src/components/views/directory/RoleDirectoryDetail.tsx:
```diff
@@ -0,0 +1,270 @@
+            email = email.substring(1);
+        }
+        let firstName = email.split('.')[0];
+        firstName = firstName.charAt(0).toUpperCase() + firstName.slice(1);
```
Please centralise this repeated logic

sppanday replied:
I have moved all of these functions into utils/ folder which is default matrix-react-sdk folder for storing sharable utility functions by their type and usage. Also took little bit to refactor one of the function to make it reusable by removing html content. Now it looks smooth. Didnt have time before but done now. All changes pushed.

jrschremmer commented on src/components/views/rooms/RoomList.tsx:
```diff
@@ -598,6 +610,12 @@ export default class RoomList extends React.PureComponent<IProps, IState> {
             }
         }
 
+        if (!config.showExplorePublicRoom) {
```
This looks odd, why is this prompt related to not showing the explore public rooms?

sppanday replied:
'use_custom_text_label_on_left_menu' has been used now with new commits pushed. Text had mentioned about explore public room which might be why I thought lets use that flag but with your suggestion, i have used above flag. With this flag any new text that is passed through config can be used. I wish matrix had provided that in beginning instead of hardcoding text on left menu panel but with my changes it re assigns text value which will align with any client's text requirement. Also documented in feature-communicate-md.

sppanday committed:
Refactored avatar logic , added description

sppanday committed:
initialized values in constructor to false,
removed redundant true/false

sppanday committed:
added new config for left panel default text
caption change to the text that client wants through config

sppanday committed:
added new text for landline extension

sppanday committed:
centralised shared functions to utility folder

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +114,42 @@ export function defaultAvatarUrlForString(s: string): string {
+    /**
+     * randomlySelectedAvatarConfigColor randomly selected color from config
+     * randomly selected color would not be used if there is no 'undefined', 'null' returned
+     * A randomly selected color would be only selected if username is not string
```
I don't see how randomlySelectedAvatarConfigColor is ever used.  In the defaultToOtherConfigColors code above, if the for loop is not entered, colorIndex will be 0 and defaultToOtherConfigColors the first configAvatarColors.

sppanday replied:
What do you recommend? I think I should hardcode 7-8 colors there and loop through and show that if config value doesnt match. Alternatively I can remove randomlySelected part. I just did not want matrix color to be applied even when name is not valid string. I think randomlySelectedColor could be used best if some colors are hardcoded so that if config colors are removed accidently then randomly select from hardcoded ones but not matrix ones.

sppanday replied:
I have removed randomly selected color as it was not good experience to see color change on avatars on every page refresh by memory. Idea was to go to our color on every possible way. I could reuse that random color to hardcode some colors if config color fails then pick color from hardcoded ones if you like. i had tested it out, pretty good but "not good part is when browser is refreshed it will randomly select a different color from index even on hardcoded one so I disliked it." To be honest, I had implemented random one before  using defaultToOtherColors so did not remove but now its been removed as there is a bit of cloud on if it will ever be used, and if it is used then user experience might be not good as users can be confused thinking why my avatar color changed suddenly.

sppanday replied:
Update: I have removed randomlyselected color entirely as per my comment above, then later I also refactored further to remove defaultToOtherConfigColor variable so when result is done then at the end just returns finalColor based on given calculations. Also updated feature-communicate.md

sppanday committed:
View renamed to adjust with service, ppl directory

sppanday committed:
renamed roleId to general queryId

sppanday committed:
Added error for directory search to display in UI

sppanday committed:
removed randomly selected color selection for
better user experience

sppanday committed:
Using 'view' instead of 'view detail'

sppanday committed:
removed defaultToOtherColor, randomlySelectedColor

sppanday committed:
aligned feature doc description to match with flag

sppanday committed:
tidied up space, text etc.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +114,46 @@ export function defaultAvatarUrlForString(s: string): string {
+     * length of 's' string passed in. For all non matching name key,
+     * select color based on name length which will never fail.
+     */
+    if (selectedAvatarColorFromConfig == undefined || !selectedAvatarColorFromConfig) {
```
I think it is safer if, this if statement, just becomes an else block on the if statement above, as if someone accidentially doesn't make this if logic mutually exclusive of the above if logic, then potentially both then blocks can be executed, or neither then blocked executed.

sppanday replied:
Done. Implemented.

jrschremmer commented on src/Avatar.ts:
```diff
@@ -112,6 +114,46 @@ export function defaultAvatarUrlForString(s: string): string {
+    /**
+     * If no color key matches form given 's' to keypair in config.json
+     * then make color stable for avatar by calculating index based on
+     * length of 's' string passed in. For all non matching name key,
```
Please updated the comment to match the implementation e.g. replace
```
     * length of 's' string passed in. For all non matching name key,
     * select color based on name length which will never fail.
```
with
```
     * content of 's' string passed in.
```

sppanday replied:
Too easy suggestion has been implemented.

jrschremmer commented on src/components/views/dialogs/InviteDialog.tsx:
```diff
@@ -199,17 +244,42 @@ interface IDMRoomTileProps {
+        let viewDetailBtn = ev.currentTarget.parentNode.querySelector<HTMLElement>("#mx_viewDetailBtn");  // selects current button clicked
+        if (detailView.style.display === "none") {
+            detailView.style.display = "block";
+            viewDetailBtn.innerHTML = "Hide Detail";
```
SHould "Hide detail" be just "Hide" to be consistent with the "View detail" changing to "View"?

sppanday replied:
I dont think it will be clear to user so I left with Hide detail there we will ask audience later. View will make sense to view item in detail but then hide detail will make user completely understand what s/he is hiding. "Hide" term may imply user is hiding entire search item so I did not make that just "hide"

sppanday replied:
View detail would sound larger text and also matrix uses just "View" for displaying room in another page from modal like for instance for "room directory"...so when view is expanded I wanted to make it clear to user you are hiding detail view by showing "hide detail". This will remove ambiguity.

sppanday committed:
added else condition on avatar color selection

sppanday committed:
changed javascript to nicer format/plain english.

sppanday committed:
refactored terms, realignes import references

sppanday committed:
replaced redundant else if with else

sppanday committed:
Capitalized Directory Search terminologies

sppanday committed:
moving categoryId check to new line

sppanday committed:
handled error when calling favorites api

sppanday committed:
handled error, simplified boolean

jrschremmer commented on src/Avatar.ts:
```diff
@@ -188,3 +229,69 @@ export function avatarUrlForRoom(room: Room, width: number, height: number, resi
+        /**
+         * If there is a problem translating then just grab first name initials as matrix default
+         */
+        if (e instanceof Error) {
```
Is `if (e instanceof Error) {` required?  I thought we would want to do `return getInitialLetter(name);` regardless of the type of e.

sppanday replied:
Removed 👍 
You are right, no need to make it conditional there anymore. I had initially used it to do console.log and see errors but not needed for final state of merging change to master.

sppanday committed:
removed instanceOf/ not needed in catch block

sppanday committed:
Moving code to newline to where it is required

sppanday committed:
refactored, centralised how avatar is named

sppanday committed:
simplified error to textual format in catch block

sppanday committed:
made search results unique

sppanday committed:
changed contact to office phone, mobile phone type

sppanday committed:
phone type refactor, removed simplifiedID from UI

sppanday committed:
aligned new server search results for roles

sppanday committed:
removed extra dash on contact

sppanday commented:
The view that calls it sets value to null if roleCategoryId is null..was already tested with matrix defaults.

jrschremmer commented:
> The view that calls it sets value to null if roleCategoryId is null..was already tested with matrix defaults.

The point is ideally we don't want to have to add the repeated new lines to the element code that call this constructor e.g.
```
                    available: false,
                    favorite: false,
                    roleCategoryId: null
```
which still exist in two callers in InviteDialog.tsx

sppanday committed:
minor change in contact displayed for directory

sppanday committed:
removed redundant data injections on Member

sppanday commented:
change has been pushed

sppanday committed:
removed redundant  fields in Member object

sppanday committed:
cleaned up phone number in enum

sppanday committed:
added missing documentation for ke formatting

sppanday committed:
reverted change as it showed some problem

sppanday committed:
role directory button and translation

sppanday committed:
minor tweak in favorite

sppanday committed:
added pager, fax, workphone text labels

jrschremmer committed:
CODE REVIEW: revert all changes to src/PageTypes.js, as no changes are required;
In Avatar.ts remove redundant code;
In HomePage.tsx, move the directory tiles above the start a chat tile to encourage users to communicate to role, as agreed in the App Playback;
In InviteDialog.tsx and DirectoryDetailView.tsx, centralise error logging to the console, and re-use an existing internationalised phrase for the error message shown to the user, instead of introducing a new hard coded English only phrase;
In InviteDialog.tsx, fix another scenario where serverResultsMixin can be set to undefined and comment out "Showing X records of Y" until it is working;
Rename config.use_custom_text_label_on_left_menu to config.left_hand_nav_help_text for clarity
In RoomList.tsx, move the config.left_hand_nav_help_text above the default conditional code for showing text in the same location, to make it obvious that config.left_hand_nav_help_text overrides any default behaviour; show roles based on config.showRoleDirectory;
In config.ts, rename use_custom_text_label_on_left_menu to left_hand_nav_help_text, for consistency and clarity;
In formatEmailUtil.ts, formatKeyValueUtil.ts, formatPhoneNumberUtil.ts and strings.ts, centralise code;
In LeftPanel.tsx, only show the Explore Public Rooms button if config.showExplorePublicRooms is true;
In directory-enums.ts, match the label to those used in the IAM application;

sppanday committed:
added regex back as role dir..avatar did not work

sppanday committed:
rename api_base_path to communicate_api_base_path

jrschremmer committed:
CODE REVIEW: In Avatar, remove the repeated regex and instead handle when getInitialLetter(lastName) is undefined;
Create DirectoryService.ts and move all directory service api calls to this class, to separate concerns, centralise api calling code and make it easier in the future to introduce mock service implementations;
Only call the search practitioner search once (before it was calling it twice: once to get the totalRecordCount and a second call to get the search results);
